taichi/codegen/cuda/codegen_cuda.cpp

```diff
@@ -147,6 +148,38 @@ class TaskCodeGenCUDA : public TaskCodeGenLLVM {
     llvm_val[stmt] = create_print(formats, types, values);
   }
 
+  void visit(AllocaStmt *stmt) override {
+    // Override shared memory codegen logic for large shared memory
+    if (stmt->ret_type->is<TensorType>() && stmt->is_shared) {
+      auto tensor_type = stmt->ret_type->cast<TensorType>();
+      size_t shared_array_threshold = 49152;
+      size_t shared_array_bytes =
+          tensor_type->get_num_elements() *
+          data_type_size(tensor_type->get_element_type());
+      if (shared_array_bytes > shared_array_threshold) {
+        if (dynamic_shared_array_bytes > 0) {
+          TI_ERROR(
+              "Only one single shared array instance is allowed in current "
+              "version.")
+        }
+        // Clear tensor shape for dynamic shared memory.
+        tensor_type->set_shape(std::vector<int>({0}));
```

Comment:
OOC: why do we need to clear tensor shape here? This is a bit weird by changing tensor shape in the codegen phase...